Improve SpikeGLX electrode metadata

CodyCBakerPhD opened the pull request:
First attempt at making the metadata for electrodes columns really explicit

How does this work for you @garrettmflynn? Or were you thinking that the column definitions in the metadata schema would also be dynamically generated for use in validation?

garrettmflynn commented:
My thought was for these property definitions to make their way into `schema.properties.Ecephys.properties.definitions.Electrodes.properties` and replace the name + description properties that are currently there. This would actually describe the shape of the data (an array of objects with keys that represent columns) and allow for descriptions to be applied for all the columns that are expected for the interface

Currently, this draft updates the `schema.properties.Ecephys.properties.Electrodes.default` property, which seems misleading to me—but could be used instead if we have an appropriate justification for this JSON Schema structure that I'm not aware of. From my understanding, `default` isn't of the expected format for a general Schema.

CodyCBakerPhD commented:
Are you expecting a different JSON definition for each entry in the schema for the `Electrodes` array?

CodyCBakerPhD commented:
Using an instance of the first 5 rows from the table generated on SpikeGLX info, can you report to me what you expect a reverse-engineered schema to look like, together with a valid instance of that schema?

garrettmflynn commented:
Since we were mainly trying to fix the column headers in the table, we just would need to update the Electrodes property definition. At present, custom columns are handled fine even without an appropriate schema definition—they just wouldn't have a description. No need to consider rows independently, they should just conform to the Electrodes object definition, which is linked with a `$ref` in the Electrodes array schema.

Here's the [full schema](https://github.com/catalystneuro/neuroconv/files/11466229/example_metadata_schema.txt) I'd expect on the frontend, including removal of the default array. It's a .txt file because it seems that GitHub doesn't let you upload JSON files.

In summary, here are the main updates I'd expect:

```json
"definitions": {
    "Electrodes": {
                        "type": "object",
                        "additionalProperties": true,
                        "required": ["channel_name", "group_name"],
                        "properties": {
                            "channel_name":{
                                "type": "string",
                                "description":  "The name of this channel."
                            },
                            "group_name": {
                                "type": "string",
                                "description":  "The name of the ElectrodeGroup this channel's electrode is a part of."
                            },
                            "contact_shapes": {
                                "type": "string",
                                "description":  "The name of the ElectrodeGroup this channel's electrode is a part of."
                            },
                            "gain_to_uV": {
                                "type": "number",
                                "description":  "The scaling factor from the data type to microVolts, applied before the offset."
                            },
                            "shank_electrode_number": {
                                "type": "integer",
                                "description": "0-based index of the electrode on the shank.",
                            },
                            "offset_to_uV": {
                                "type": "number",
                                "description":  "The offset from the data type to microVolts, applied after the gain."
                            },
                            "inter_sample_shift":{
                                "type": "number",
                                "description":  "Time-delay of each channel sampling in proportion to the per-frame sampling period."
                            }
                        }
                    }
                }
```

Let me know if anything remains unclear.

CodyCBakerPhD commented:
Thanks that helps clarify a lot.

A couple questions though...

1. How is it validated that an `addditionalProperty` in an instance of this schema follows the `name`+`description`+`type` form?

2. Given this is an object now, not an array, how do we specify the order of the columns (which is something we actually would want to control at the API level)?

3. The actual data type of each property can matter a lot (`uint8` vs. `int64`, I guess less so for `float32` vs. `float64` but still...) but the encoding mapping that schematically maps the column `type` to JSON types is many-to-one, so they cannot be decoded without someone recording the information about what data type is was before JSON encoding. That's why I included a `data_type: str` field in my current definition; how would that fit into your proposal?

CodyCBakerPhD commented:
3. In reference to your .txt file of the complete schema, would that actually work? I'm confused how an `array` can have `minItems` of zero, each item following a definition, and that definition being an object with multiple requirements

Basically, what would an instance of the object look like with custom columns added to it?

garrettmflynn commented:
> How is it validated that an additionalProperty in an instance of this schema follows the name+description+type form?

As far as my experience with JSON Schema goes, this is just what the standard allows (where the name is actually the property key). Was there something else you meant by this?

When adding new properties, I assumed we wouldn't specify anything about them per-se—though this could mean we wouldn't have the right input option. Would we want to start passing a modified schema back to neuroconv so it would have access to this?


> The actual data type of each property can matter a lot (uint8 vs. int64, I guess less so for float32 vs. float64 but still...) but the encoding mapping that schematically maps the column type to JSON types is many-to-one, so they cannot be decoded without someone recording the information about what data type is was before JSON encoding. That's why I included a data_type: str field in my current definition; how would that fit into your proposal?

Oh sure. My proposal was simply trying to fit everything into JSON Schema. Since everything would end up passing through a JSON parser to get from JavaScript back to Python anyways (thus obscuring the original data type of any input), couldn't we just parse the passed data based on the expected / actual type? 

AFAIK this consideration would make it unnecessary to have those original types on the frontend. Does this make sense? Or are there consideration we'd have to make for new properties?

> In reference to your .txt file of the complete schema, would that actually work? I'm confused how an array can have minItems of zero, each item following a definition, and that definition being an object with multiple requirements. Basically, what would an instance of the object look like with custom columns added to it?

```js
const emptyElectrodesTable = [] // Empty electrode table (would be minItems: 0)

 // This would be the result to return. It conforms with the object definition with multiple requirements
const singleElectrodeTable = [
   {
      channel_name: 'test',
      group_name: 'group',
      custom_property: True
   }
]

const invalid_electrode = {
    custom_property: True
} // Missing required properties. Would fail if added to the table
```

Does this make sense?

CodyCBakerPhD commented:
> As far as my experience with JSON Schema goes, this is just what the standard allows (where the name is actually the property key). Was there something else you meant by this?

> When adding new properties, I assumed we wouldn't specify anything about them per-se—though this could mean we wouldn't have the right input option. Would we want to start passing a modified schema back to neuroconv so it would have access to this?

The way it works now is the definition of `Ecephys/Electrodes` is just an array (possibly an empty one) where every element is a `name`+`description` defining the `name` and `description` of a column on the electrodes table

So I can pass in name/description pairs for predefined columns or custom ones and they both follow the same definition

Where do the 'test' and 'group' values in your example end up on the table at the end of the day? Are those actual values for an actual electrode? Are those the descriptions of the column whose name is the key?


> AFAIK this consideration would make it unnecessary to have those original types on the frontend. Does this make sense? Or are there consideration we'd have to make for new properties?

When adding a custom column  with custom data that would restrict the user to either `int` or `float` from `integer` or `numeric`; might be best to let the user define what kind of data type they'd like to use instead of assuming


What might be the thing that is confusing here is how we have two uses for the term `description`; this is an actual value in an instance of the JSON object that ends up as an actual value set on the column in the NWB file, not just a default to use to annotate that part of the JSON schema

This results in the metadata structure we currently have, which results in JSON object instances like

```
{
    Ecephys: {
        Electrodes: [
            {
                name: "channel_name",
                description: "The name of this channel."
            },
            {
                name: "custom_channel",
                description: "My custom description for my custom channel, which will end up the NWB file itself."
            }
        ]
    }
}
```

Really, this field of the schema should be called `ElectrodeColumn` instead of `Electrodes` since it doesn't actually specify information about the electrodes themselves, just what properties they can have

The schema definition of an `ElectrodeColumn` is that is has a `name` that can be set, a `description` that can be set, and I suppose a data type that can also be set (not currently in the schema but that's what this PR includes too)

But if I expand upon your example, the JSON schema defined in your shared .txt file would actually produce the following as a possibility

```
{
    Ecephys: {
        Electrodes: [
            {
                 channel_name: 'test',
                 group_name: 'group',
            },
            {
                custom_property: ""
            }  # < - minItems = 0 means I should be able to add more objects to this
               # but I guess the definition means they'd each have to have a 'channel_name' and 'group_name' key?
               # But there can only be one electrode table in the file and those columns are already defined
        ]
    }
}
```

CodyCBakerPhD commented:
Summary of conclusions from discussion:

- the GUIDE table is actually two tables in one entity
  - the header is the NeuroConv metadata which is validated by NeuroConv metadata schema
    - This needs the true original `data_type` of the SI property to be recorded for roundtrip decoding; added in this PR but needs some adjustment in lower level asserts
  - the body of the table (the rows of electrodes) are validated by a separate schema that does not really exist yet; this is easily added however by harnessing `pandas` which can read the JSON form of the previous helper function and already has utils for automatically building a JSON schema using the [JSON table specification](http://dataprotocols.org/json-table-schema/); added in this PR, but need to add some tests for it

CodyCBakerPhD commented:
I'll resolve the testing stuff on this PR next week, otherwise the new methods should be good to test on the GUIDE @garrettmflynn

CodyCBakerPhD commented:
... though the next step that is required is to add some helper functions that accept a modified JSON object and add/update the property values in the recording extractor

But I think we were still waiting on the 'add column' table functionality for that anyway as I recall

CodyCBakerPhD commented on src/neuroconv/tools/spikeinterface/spikeinterface.py:
```diff
@@ -215,12 +215,12 @@ def add_electrodes(recording: BaseRecording, nwbfile: pynwb.NWBFile, metadata: d
         electrodes_metadata = metadata.get("Ecephys", dict()).get("Electrodes", list())
 
-    required_keys = {"name", "description"}
+    required_keys = {"name", "description", "data_type"}
```
Yeah I was aware of that; for the Huszar conversion specifically, use `main`

Currently this PR is only for current dev branch of the GUIDE; a bit asynchronous for the moment I'll admit but will try to fix this and get it through soon

garrettmflynn replied:
Perfect just wanted to make sure it didn't slip through

CodyCBakerPhD commented:
This approach has been abandoned in favor of GUIDE specific functionalities
